[RFC] tests: Migrate legacy test 61.

lucc opened the pull request:
There are a lot of fundamental differences from the legacy test in this:
- the file test.out and the final expect() call is not used.  This was neccessary b/c the old test used `.w >>test.out` a lot which did not seem to work in the new test. So I decided to `eq('text', eval('getline(".")'))` instead.
- the old test uses `set ul=100` a lot.  This is not needed because the `feed()` and `execute()` functions simulate real user interaction.  The old test did not.

Please comment if you think this is the correct or incorrect way to address it.

This test was causing some travis failures recently so I decided to do it despite the complications.

Sorry for the messy commits. I will sort them on the way and squash at the end.

lucc committed:
tests: Migrate legacy test 61.

This test is real time based as it also tests the `:earlier` and `:later`
commands with time arguments (using `:sleep`).  This can sometimes case the
test to fail on systems that are under heavy load or where the time interval
between creating the expected buffer state and the `:earlier` or `:later`
command that tries to jump to it changes.

To be system independent we use nvim's `:sleep` command and `wait()` for it in
the test suit.

The legacy vim test writes to test.out a lot with `:.w >>test.out`.  This
does currently not work in the lua test suite so the test is modernized to use
busted's assertions instead of the output file.

This test was treated special in the legacy Makefile but after the conversion
the related code can be removed.

lucc committed:
tests: Don't set ul in migrated test 61.

The legacy test uses `set ul=100` to break the changes into blocks that can be
undone separately.  This is needed because the legacy test is sourced from a
file and changes would be grouped into on undo block by default.  The lua test
suite does not have this restriction.

Also add a new test case to test this effect of using `set ul=100` in a
sourced script.

justinmk commented:
>  the old test used `.w >>test.out` a lot 

This could be the reason test61 fails randomly. We've seen similar fsync-related issues: https://github.com/neovim/neovim/pull/1678#issuecomment-67234613

Thanks for tackling this!

lucc commented:
currently the semantics depart in two points. Both are marked `TODO` in the code.

All other checks are done. I went trough `test61.ok` line by line to see that I check the same stuff. The legacy tests does nearly all of them with `.w >>test.out` and `%w >>test.out`. The new test uses `eq('text', eval('getline(".")')` and `expect([[some text]])` instead.

lucc commented:
I updated the commit history. It is now slightly better suited to understand what is happening. I will still squash this later and write a summarized commit message with info from the other commit messages and the discussion here.

lucc commented on test/functional/legacy/061_undo_tree_spec.lua:
```diff
@@ -0,0 +1,186 @@
+    feed('occcc<esc>u')
+    -- TODO At this point the original test will write "aaaa" to test.out.
+    -- Why is the line "bbbb" here?
+    eq('bbbb', eval('getline(".")'))
```
The behavior in vim and nvim if I perform this step manually is the same as in this new test file. The two changes (`obbbb<esc>` and `occcc<esc>`) are not joined and can be undone separately.

lucc commented:
I found some more things about the problem in line 81-90. The description of `:undojoin` is not the best:
```
:undoj[oin]		Join further changes with the previous undo block.
			Warning: Use with care, it may prevent the user from
			properly undoing changes.  Don't use this after undo
			or redo.

This is most useful when you need to prompt the user halfway through a change.
For example in a function that calls |getchar()|.  Do make sure that there was
a related change before this that you must join with.

This doesn't work by itself, because the next key press will start a new
change again.  But you can do something like this: >

	:undojoin | delete

After this an "u" command will undo the delete command and the previous
change.
```
But it seems to me that this is only works in scripts and such (command chains or inside `:norm`). The hint might be in the example about `undoj|d`. If I do the following in a interactive session (typed keys including newline expect for <esc> wich should be real esc char):
```
oaaa<esc>obbb<esc>:undoj
occc<esc>u
```
I end up with three undo blocks. The same is true if I source the following script: 
```vim
normal oaaaa
set ul=100
normal obbbb
set ul=100
normal occcc
```
(if I leave out the `set ul=100` I end up with only one change)
but here (in the script) `undoj` works like one expects (first change = "\naaa", second change = "\nbbb\nccc"):
```vim
normal oaaaa
set ul=100
normal obbbb
set ul=100
undojoin
normal occcc
```
I will adjust the test file accordingly.

lucc committed:
tests: Fix an further simplify migrated test 61.

:undojoin can only be used inside scripts and command chains.  So it has to be
tested inside an explicit `source()` call.

Also add a new test case for the different behavior when sourceing normal mode
commands from a script or inserting them interactively.

lucc commented:
I think we should also update the help text for `:undojoin` and `'ul'` to explain this strange behavior better. Should I add a commit to this PR or open a new one for it?

justinmk commented:
> I think we should also update the help text for :undojoin and 'ul' to explain this strange behavior better. Should I add a commit to this PR or open a new one for it?

That would be better served by an email to vim_dev.

justinmk commented:
> If I do the following in a interactive session (typed keys including newline expect for wich should be real esc char):
> `oaaa<esc>obbb<esc>:undoj`
> `occc<esc>u`

Yeah, that's what the help means when it says "by itself". I guess vim_dev might say that is good enough. If you can think of an extra sentence or rephrasing, we can just do that in this PR.

lucc commented:
>> If I do the following in a interactive session (typed keys including newline expect for wich should be real esc char):
>> `oaaa<esc>obbb<esc>:undoj`
>> `occc<esc>u`

> Yeah, that's what the help means when it says "by itself". I guess vim_dev might say that is good enough. If you can think of an extra sentence or rephrasing, we can just do that in this PR.

But then again if I save these characters in a file and run it with `:so!` I get the behavior from the legacy test. I now added both versions to the test (see upcoming commit).

lucc commented:
This is ready for review.

justinmk commented:
>  if I save these characters in a file and run it with :so! I get the behavior from the legacy test.

Because `:so` enters a non-interactive state. Vim behavior very strongly depends on this. If a sequence is entered at the `:` command-line, Vim behaves differently in several ways (one example being the logic of the "press enter" prompt).

justinmk commented on src/nvim/testdir/Makefile:
```diff
@@ -126,15 +126,6 @@ test1.out: .gdbinit test1.in
-	                 else rm -rf $*.failed $(RM_ON_RUN);  \
-	                     $(RUN_VIM) $*.in;                \
-	                 fi;                                  \
-	             fi"
```
:+1:  I forgot that even the Vim Makefile gives special attention to test61.

justinmk commented on test/functional/legacy/061_undo_tree_spec.lua:
```diff
@@ -0,0 +1,256 @@
+-- Tests for undo tree and :eralier and :later.
```
typo

justinmk commented on test/functional/legacy/061_undo_tree_spec.lua:
```diff
@@ -0,0 +1,256 @@
+end
+local expect_line = function(line)
+  return eq(line, eval('getline(".")'))
+end
```
:+1:  good compromise (hiding these workarounds in `expect_foo` functions)

justinmk commented:
Just for fun I'm going to restart the build a couple times to see how reliable it is now :)

justinmk commented on test/functional/legacy/061_undo_tree_spec.lua:
```diff
@@ -0,0 +1,256 @@
+    expect_line('123456')
+
+    -- Delay for three seconds and go some seconds forward and backward.
+    execute('sleep 2')
```
I guess it's worth updating the comment to "two seconds". Seems like the original comment was outdated.

justinmk commented on test/functional/legacy/061_undo_tree_spec.lua:
```diff
@@ -0,0 +1,256 @@
+  helpers.feed, helpers.insert, helpers.source, helpers.eq, helpers.eval,
+  helpers.clear, helpers.execute, helpers.expect, helpers.wait
+
+local expect_empty_buffer = function()
```
Note that we prefer the `local function foo()` syntax instead of `local foo = function()`. (Mostly because we just need to pick one and stick with it)

justinmk commented on test/functional/legacy/061_undo_tree_spec.lua:
```diff
@@ -0,0 +1,256 @@
+    expect_line('123456abc')
+
+    -- Test that setting 'ul' breaks change blocks, we need to use source() in
+    -- order to test this, as interactive changes are not grouped.
```
This block (up to line 131) should be a separate `it()` if it's not too much trouble.

justinmk commented:
I restarted the travis build twice (for a total of 3 full travis runs). The first 2 passed all builds. The 3rd run [failed with  test61 error](https://travis-ci.org/neovim/neovim/jobs/67473051) on the OSX clang build:

```
441 successes / 1 failure / 0 errors / 1 pending : 41.185182 seconds
Pending -> ...vis/build/neovim/neovim/test/functional/job/job_spec.lua @ 328
jobs will only emit the "exit" event after "stdout" and "stderr"
Failure -> ...vim/neovim/test/functional/legacy/061_undo_tree_spec.lua @ 30
the undo tree is working
...vim/neovim/test/functional/legacy/061_undo_tree_spec.lua:89: Expected objects to be the same.
Passed in:
(string) '123456a'
Expected:
(string) '123456'
stack traceback:
	...vim/neovim/test/functional/legacy/061_undo_tree_spec.lua:89: in function <...vim/neovim/test/functional/legacy/061_undo_tree_spec.lua:30>
-- Output to stderr:
CMake Error at /Users/travis/build/neovim/neovim/cmake/RunTests.cmake:39 (message):
  Running functional tests failed with error: 1.
```

justinmk commented on test/functional/legacy/061_undo_tree_spec.lua:
```diff
@@ -0,0 +1,256 @@
+    feed('Ab<esc>')
+    feed('Ac<esc>')
+    expect_line('123456abc')
+    execute('ear 1s')
```
I'm just guessing, but maybe add a `wait()` call after these `:earlier` commands to see if that helps with the random test failures. 

Also, I suggest spelling out the full command (`:earlier` rather than `:ear`) unless there's some space constraints that make it onerous to do so.

justinmk commented on test/functional/legacy/061_undo_tree_spec.lua:
```diff
@@ -0,0 +1,256 @@
+      c
+      12]])
+    feed('od<esc>')
+    -- The file Xtest.source is write during setup.  It contains this text
```
"is write" => "is written"?

justinmk commented on test/functional/legacy/061_undo_tree_spec.lua:
```diff
@@ -0,0 +1,256 @@
+      12
+      d]])
+    -- The above behaviour was tested in the legacy vim test because the
+    -- legacy tests where executed with ':so!'.  The behavior differs for
```
where => were

justinmk commented:
@lucc I compared your test to the old one. Great work. Also, I understand now why this test is flaky: it is inherently time-based.

lucc committed:
tests: Split test 61 into several it() blocks.

lucc commented:
@justinmk I applied your comments. I did also keep the commit about 'ul' but I added some related changes from another commit to it.

I also split the test into several smaller tests.

Is there anything that can be done about the failures that occur due to the real time reliance of the test?

lucc commented:
Is there a way to specify a different busted output format which makes it is easier to grep for errors and such? I am running something like 
`busted --filter=time --repeat=100 $TEST_FILE` but the output is huge, sparse and not nice to grep.

Best would be not output and a small report about the number of failures at the end of all repetitions.

justinmk commented on test/functional/legacy/061_undo_tree_spec.lua:
```diff
@@ -0,0 +1,260 @@
+describe('undo:', function()
+  before_each(clear)
+  teardown(function()
+    os.remove('Xtest.source')
```
Since this is removed in a teardown(), would be a bit clearer to create it setup().

lucc replied:
I would prefer to remove the teardown then. The file is best created in the it() block because I reuse the contents. compare line 181-183 and 239.

justinmk commented:
> Is there anything that can be done about the failures that occur due to the real time reliance of the test?

My current thought is to move the `:earlier` and `:later` `it()` tests into local functions, then retry them up to 3 times.

I don't know if it's possible to "catch" a failed expect/assert (e.g. with [pcall](http://www.lua.org/pil/8.4.html))--if that's possible it might be the path of least resistance, though maybe a bit ugly. If that doesn't work then I guess the `expect()` calls would need to be removed and instead write a local function like `try_expect` which returns `{false, "<failure message>"}` which is then retried up to 3 times by the actual `it()` block. If it fails 3 times then the `it()` will finally `assert()` with the relevant message.

Does that make sense?

lucc commented:
We don't need to catch any expects/asserts, we can just check for equality ourself in the local function. The big problem with this is that the test sleeps for three seconds. I consider that to quite long. But if we start it several times this only increases. So I'm not really for it.

The problem is not worse than with the legacy test so one (maybe temporary) solution might be to add a note to the test or error message and leave it as is.

justinmk commented:
> Is there a way to specify a different busted output format which makes it is easier to grep for errors and such?

Kind of. You can specify different [output handlers](https://github.com/Olivine-Labs/busted/tree/master/busted/outputHandlers):

    busted --filter=time --repeat=100 -o gtest $TEST_FILE 

Here's the relevant line in CMakeLists.txt:

https://github.com/neovim/neovim/blob/master/CMakeLists.txt#L256

Note that you can edit the output handler `.lua` code directly in your local `.deps/` if none of them give the minimized output you're looking for.

justinmk commented:
> We don't need to catch any expects/asserts, we can just check for equality ourself in the local function. 

Right, that was my 2nd suggestion in case it wasn't clear. 

> The big problem with this is that the test sleeps for three seconds. I consider that to quite long. But if we start it several times this only increases.

We only restart it if it fails, so the test make take 3, 6, or 9 seconds. Avoiding a false failure saves us a *lot* of time, by not having to restart spurious build failures.

lucc commented:
ok will implement it.

lucc committed:
tests: Repeat a flaky part of the migrated test 61.

This is the part of the test that relies on wall clock time and sometimes
fails if the system is under load.  The test is repeated up to three times
before a failure is reported to the user.

lucc commented:
pcall works! Do you have an idea about getting the original error message to show up for the user (see TODO in code)?

lucc commented:
A test failure can be simulted for the relevant test by replacing `return` with `break` in line 108.

justinmk commented:
> Do you have an idea about getting the original error message to show up for the user

Don't use `pcall` for the last attempt :)

lucc commented:
@justinmk nice idea. So what do you think about the comments and notes? How often should we retry the test overall? Any notes for the squashed commit message?

justinmk commented:
> How often should we retry the test overall?

3 is good for now. We will need to monitor it for a couple weeks.

justinmk commented on test/functional/legacy/061_undo_tree_spec.lua:
```diff
@@ -0,0 +1,280 @@
+	-- Assert that no undo history is present.
+	eq({}, eval('undotree().entries'))
+	-- Delete three characters and undo.
+	feed('Gxxx')
```
spaces, not tabs.

justinmk commented on test/functional/legacy/061_undo_tree_spec.lua:
```diff
@@ -0,0 +1,280 @@
+      end
+      -- We did not return in the loop so there was an error or failure.
+      -- We now try to run the test again but will not catch further errors,
+      -- so the user will see them.
```
can just remove this comment and instead mention above line 104:

    Retry up to 3 times.

justinmk commented on test/functional/legacy/061_undo_tree_spec.lua:
```diff
@@ -0,0 +1,280 @@
+      -- encapsuled in a function and will be tried up to three times if it
+      -- fails.  This is done because the test is dependent on wall clock time
+      -- and somtimes fails if the system is under load or the test is paused
+      -- for a long enough interval.
```
Instead of this comment, consider renaming `test()` to `test_earlier_later()` with this comment:

```
-- `:earlier` and `:later` are (obviously) time-sensitive, so this test
-- sometimes fails if the system is under load. It is wrapped in a local 
-- function to allow multiple attempts.
```

lucc commented:
Applied comments. 

I also added `clear() at the top of `test_earlier_later()` because the normal `before_each()` setup will not be redone for a second execution.

justinmk commented:
Nice. Great contribution, thank you. Will merge after build.

justinmk commented:
Hmm. OSX builds timed out, but other builds are fine. Restarted all travis builds.
